File: C++/breadth-first_search/breadth-first_search.cpp
//
// Created by hunor on 24.05.2023.
//

#include "breadth-first_search.h"

Graph::Graph(int number_of_nodes) : number_of_nodes(number_of_nodes) {
    adjacency_list.resize(number_of_nodes + 1); // Adjust for 1-based indexing
}

void Graph::add_edge_directed(int first_node, int second_node) {
    adjacency_list[first_node].push_back(second_node);
}

void Graph::add_edge_nondirected(int first_node, int second_node) {
    adjacency_list[first_node].push_back(second_node);
    adjacency_list[second_node].push_back(first_node);
}

void Graph::breadth_first_search(int start_node) {
    vector<bool> visited(number_of_nodes + 1, false); // Track visited nodes
    queue<int> node_queue;
    node_queue.push(start_node);
    visited[start_node] = true;

    while (!node_queue.empty()) {
        int current_node = node_queue.front();
        cout << current_node << " ";
        node_queue.pop();
        for (auto neighbor : adjacency_list[current_node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                node_queue.push(neighbor);
            }
        }
    }
    cout << endl;
}

File: TypeScript/src/breadth-first-search/breadth-first-search.ts
import { Queue } from 'queue-typescript'; // Import a queue library for TypeScript

/**
 * Represents a class for performing Breadth-First Search (BFS) on a graph.
 */
export class BFSGraph {
  private numberOfNodes: number; // The number of nodes in the graph
  private adjacencyList: number[][]; // Adjacency list representing the graph

  /**
   * Constructor for the BFSGraph class.
   * @param numberOfNodes The number of nodes in the graph.
   */
  constructor(numberOfNodes: number) {
    this.numberOfNodes = numberOfNodes;
    this.adjacencyList = new Array<number[]>(numberOfNodes + 1).fill([]);
  }

  /**
   * Add an undirected edge between two nodes in the graph.
   * @param firstNode The first node.
   * @param secondNode The second node.
   */
  addEdgeNonDirected(firstNode: number, secondNode: number): void {
    this.adjacencyList[firstNode].push(secondNode);
    this.adjacencyList[secondNode].push(firstNode);
  }

  /**
   * Add a directed edge from the first node to the second node in the graph.
   * @param firstNode The source node.
   * @param secondNode The target node.
   */
  addEdgeDirected(firstNode: number, secondNode: number): void {
    this.adjacencyList[firstNode].push(secondNode);
  }

  /**
   * Perform Breadth-First Search (BFS) starting from a specified node.
   * @param startNode The starting node for BFS traversal.
   */
  breadthFirstSearch(startNode: number): void {
    const visited: boolean[] = new Array<boolean>(this.numberOfNodes + 1).fill(false);
    const nodeQueue: Queue<number> = new Queue<number>();
    nodeQueue.enqueue(startNode);
    visited[startNode] = true;

    const traversedNodes: number[] = [];

    while (nodeQueue.length > 0) {
      const currentNode: number = nodeQueue.dequeue() as number;
      traversedNodes.push(currentNode);
      for (const neighbor of this.adjacencyList[currentNode]) {
        if (!visited[neighbor]) {
          visited[neighbor] = true;
          nodeQueue.enqueue(neighbor);
        }
      }
    }

    console.log(traversedNodes.join(' '));
  }
}
